### PR TITLE
Add jwt assertion handler'

### PR DESCRIPTION
**Pull request by mcygoogle**
Add jwt assertion handler and tests

### CODE REVIEW

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwkStore.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.UUID;
+import net.minidev.json.JSONArray;
+import net.minidev.json.JSONObject;
+
+/** Store key for signing */
+@Singleton
+public class JwkStore {
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwkStore.java:32]: Is this used for risc?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
+      throws OAuth2Exception {
+    Jws<Claims> jws;
+
+    String email = null;
+    String googleAccountId = null;
+    try {
+      jws = Jwts.parserBuilder().setSigningKeyResolver(keys).build().parseClaimsJws(assertion);
+      if (!jws.getBody().getIssuer().equals("https://accounts.google.com")) {
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:107]: Use constant

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        throw new InvalidRequestException(ErrorCode.INVALID_JWT_ISS);
+      }
+      if (!jws.getBody().getAudience().equals(aud)) {
+        throw new InvalidRequestException(ErrorCode.WRONG_JWT_AUD);
+      }
+      email = jws.getBody().get("email", String.class);
+      googleAccountId = jws.getBody().get("sub", Long.class).toString();
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:114]: Constant for "email" and "sub"

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    } catch (JwtException ex) {
+      log.log(Level.INFO, "JWT Decode ERROR!", ex);
+      throw new InvalidRequestException(ErrorCode.INVALID_JWT);
+    }
+    return ImmutableList.of(email, googleAccountId);
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:119]: You can use an inner class with [@AutoValue](https://github.com/google/auto/tree/master/value) or at least Pair<String, String>

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      case CREATE:
+        handleCreateAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:99]: Can this be private?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    }
+  }
+
+  public void handleCreateAssertion(
+      HttpServletResponse response, String email, String googleAccountId, List<String> scopes)
+      throws IOException, InvalidScopeException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    if (user.isPresent()) {
+      returnLinkError(response, email);
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:170]: Is there any place mentioned we shall return error when user exists?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    } else {
+      UserDetails newUser =
+          UserDetails.newBuilder()
+              .setUsername("GAL:" + email)
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:174]: why **"GAL" +** ?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    response.getWriter().flush();
+  }
+
+  public void handleGetAssertion(
+      HttpServletResponse response, String email, String googleAccountId, List<String> scopes)
+      throws IOException, InvalidScopeException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    if (user.isPresent()) {
+      ClientDetails client = clientDetailsService.getClientByID("google").get();
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:146]: Shouldn't the client id be passed in and verified with client secret? Same for other place

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwkStore.java (hunk 1 above)
Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwkStore.java:32]: Yes, but I need it for tests here.

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 5 above)
Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:99]: I set it to public for tests.

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 6 above)
Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:170]: Yes.
Check whether either of the following conditions are true:

The Google Account ID, found in the assertion's sub field, is in your user database.
The email address in the assertion matches a user in your user database.
If either condition is true, prompt the user to link their existing account with their Google Account. To do so, respond to the request with an HTTP 401 error that specifies error=linking_error and gives the user's email address as the login_hint.

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 7 above)
Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:174]: just want to set a unique username.

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 8 above)
Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:146]: Actually from the description of Assertion, I do not see client id and secret.
I let the filter set google's client for it.
I can read the client from session here but they are actually the same.

Comment (weiminhuang-corp) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:146]: It's not in the assertion content, but either in the header or body.

Comment (mcygoogle) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java]: But in https://developers.google.com/identity/account-linking/oauth-with-sign-in-linking?oauth=code,
I do not see that. The sample is:
POST /token HTTP/1.1
Host: oauth2.example.com
Content-Type: application/x-www-form-urlencoded

grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&intent=check&assertion=JWT&scope=SCOPES

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            googleAccountId,
+            oauth2Request.getRequestBody().getScopesList(),
+            client);
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  public Pair<String, String> verifyAndGetInfoFromJwt(
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:112]: Can this be private?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+      email = jws.getBody().get(EMAIL, String.class);
+      googleAccountId = jws.getBody().get(SUB, String.class);
+    } catch (JwtException ex) {
+      log.log(Level.INFO, "JWT Decode ERROR!", ex);
+      throw new InvalidRequestException(ErrorCode.INVALID_JWT);
+    }
+    return Pair.of(email, googleAccountId);
+  }
+
+  public void handleCheckAssertion(
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:135]: private?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    } else {
+      json.appendField("account_found", "false");
+      response.setContentType("application/json;charset=UTF-8");
+      response.getWriter().println(json.toJSONString());
+      response.setStatus(HttpStatus.SC_NOT_FOUND);
+    }
+    response.getWriter().flush();
+  }
+
+  public void handleGetAssertion(
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:153]: private?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        scopes = client.getScopesList();
+      }
+      returnToken(response, client, user.get(), scopes);
+
+    } else {
+      returnLinkError(response, email);
+    }
+  }
+
+  public void handleCreateAssertion(
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:180]: private?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 7 above)
Comment (weiminhuang-corp) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:194]: This makes it very specific for Google, what about clientId(the one you give clients)+email?

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 8 above)
Comment (weiminhuang-corp) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java]: For your case, it should be in post parameter.

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 6 above)
Comment (weiminhuang-corp) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java:190]: SG

File: src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java (hunk 5 above)
Comment (weiminhuang-corp) [reply src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java]: Then you can make it package private and annotate as @VisibleForTesting

**Pull request merged**